File: array/greedy/max_product_subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maxx = float('-inf')
        prod = 1

        for v in nums:
            prod = prod * v
            if prod > maxx:
                maxx = prod
            if prod == 0:
                prod = 1
        
        prod = 1

        for v in nums[::-1]:
            prod = prod * v
            if prod > maxx:
                maxx = prod
            if prod == 0:
                prod = 1
        return maxx

File: array/greedy/max_product_subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let max = -Infinity;
    let prod = 1

    for (let v of nums){
        prod = prod * v
        if(prod > max){
            max = prod
        }
        if(prod == 0){
            prod = 1
        }
    }

    prod = 1
    for (let i = nums.length -1 ; i>=0 ; i--){
        prod = prod * nums[i]
        if(prod > max){
            max = prod
        }
        if(prod == 0){
            prod = 1
        }
    }

    return max
};